### PR TITLE
[b/323519703] Fix Oracle XML tasks

### PR DESCRIPTION
**Pull request by kpiotrowski**
b/323519703
Existing Oracle XML tasks fail due to the missing argument placeholders in XML SQL query. What's more the XML metadata is retrieved sequentially for each object what makes it extremely slow. This PR fixes XML related issues and reduces extraction time:

- Replace custom XML task with regular SQL query to improve the performance. 
- Add filters for *_TABLES and *_TASKS to exclude objects not existing in DBMS_METADATA.

Extraction time was reduced from over 10h to ~25 minutes.
XML Metadata is correctly extracted for all objects.

image

### CODE REVIEW

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/oracle/OracleMetadataConnector.java [hunk 1]
```diff
@@ -237,36 +127,57 @@ private static SelectTask newSelectStarTask(
...
-        sb.append("('").append(owner).append("'");
-      } else {
-        sb.append(",'").append(owner).append("'");
-      }
-    }
-    sb.append(")");
-    return sb.toString();
+    return (owners == null || owners.isEmpty())
+        ? null
+        : String.format("('%s')", String.join("','", owners));
```
Comment (dawidxc) [inline dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/oracle/OracleMetadataConnector.java:176]: I would recommend replacing `String.format` with concatenation:

```
"('" + String.join("','", owners) + "')"
```

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/oracle/OracleMetadataConnector.java [hunk 2]
```diff
@@ -198,22 +91,19 @@ public MessageTask(@Nonnull GroupTask... ts) {
...
+      for (int i = 1; i <= tasks.length; i++) {
         LOG.error(
-            "("
-                + c
-                + "): "
-                + task.getName()
-                + " : "
-                + ExceptionUtils.getRootCauseMessage(task.getException()));
-        c += 1;
+            String.format(
```
Comment (dawidxc) [inline dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/oracle/OracleMetadataConnector.java:97]: I would suggest avoiding `String.format` in logger invocations. The logger has built-in parameter handling.

```
LOG.error("({}): {} : {}", i, tasks[i].getName(), ...);
```

**Review by dawidxc — APPROVED** [review]

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/oracle/OracleMetadataConnector.java [hunk 3]
```diff
@@ -198,22 +91,18 @@ public MessageTask(@Nonnull GroupTask... ts) {
     @Override
     protected Void doRun(TaskRunContext context, @Nonnull ByteSink sink, @Nonnull Handle handle)
         throws Exception {
-      int c = 1;
       LOG.error("All the select tasks failed:");
-      for (GroupTask task : tasks) {
+      for (int i = 1; i <= tasks.length; i++) {
```
Comment (shevek-google) [inline dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/oracle/OracleMetadataConnector.java:95]: Is this loop correct? We index onto the tasks array, which should go from 0 to i < tasks.length, not 1?

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/oracle/OracleMetadataConnector.java [hunk 4]
```diff
@@ -198,22 +91,18 @@ public MessageTask(@Nonnull GroupTask... ts) {
...
-                + c
-                + "): "
-                + task.getName()
-                + " : "
-                + ExceptionUtils.getRootCauseMessage(task.getException()));
-        c += 1;
+            "({}): {} : {}",
+            i,
+            tasks[i].getName(),
+            ExceptionUtils.getRootCauseMessage(tasks[i].getException()));
```
Comment (shevek-google) [inline dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/oracle/OracleMetadataConnector.java:100]: These index operations make me nervous.

**Review by shevek-google — DISMISSED** [review]

File: dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/oracle/OracleMetadataConnector.java (hunk 3 above)
Comment (kpiotrowski) [reply dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/oracle/OracleMetadataConnector.java]: Thanks for spotting this bug, this is fixed now. I removed array index operations too.

**Comment by kpiotrowski** [comment]
@shevek-google Would you have some time to re-review this PR?

**Pull request merged**